### ISSUE TITLE
Error: 0x80040305

### ISSUE THREAD

**Issue by BtheDestroyer** [Microsoft/WSL#3001]
* Your Windows build number: 10.0.17604.1000

* What you're doing and what's happening: I noticed that my default user had changed to root (I don't remember if I had done that or if I had just never noticed it before. I ran `cut -d: -f1 /etc/passwd` to get a list of all users and saw "bryce" at the bottom. I wanted my user to be "bthedestroyer" so I ran `usermod -l bthedestroyer bryce` and checked /etc/passwd again to find it changed. I then closed ubuntu and ran `ubuntu config --default-user bthedestroyer` from the command line and got a confirmation message, but when I went back in I was still root. I then restarted my computer and when I tried to run ubuntu this time, I get the error 0x80040305 and it instantly closes. I tried changing the default user back to root hoping that'd fix the issue but now I get `Failed to set default user to: root`

* What's wrong / what should be happening instead: I should be able to open Ubuntu or at least change the user back to root.

[lxcore_kernel.zip](https://github.com/Microsoft/WSL/files/1787888/lxcore_kernel.zip)

**Comment by benhillis** [comment]
@BtheDestroyer - When you attempted to change the default UID back to root was there an error?

**Comment by BtheDestroyer** [comment]
I agree. The only error I got was Failed to set default user to: root

**Comment by therealkenc** [comment]
> I think we should relax this error path

That or add a `--userspec` commandline parameter. Which get you another step [closer](http://man7.org/linux/man-pages/man1/chroot.1.html).

**Comment by benhillis** [comment]
@BtheDestroyer - I'm not sure where that error message is coming from, I guess ubuntu.exe's code.

**Comment by BtheDestroyer** [comment]
My solution to get Ubuntu back running again was to copy out the files I needed from `C:\Users\%USERNAME%\AppData\Local\Packages\CanonicalGroupLimited.UbuntuonWindows_79rhkp1fndgsc\LocalState\rootfs`, uninstall Ubuntu, reinstall it, and copy the files back in. Works, but this probably shouldn't be something that happens in the first place.

**Comment by benhillis** [comment]
Agreed. I submitted a mitigation for this yesterday which will fall back to root if there is a problem with parsing the /etc/passwd file.

**Comment by Biswa96** [comment]
Can bash run without `/etc/passwd`?

**Comment by benhillis** [comment]
Yes, but the way WSL sets up environment uses /etc/passwd similar to the login binary. After my change a missing passwd file won't prevent launching.

**Comment by Biswa96** [comment]
Can you share what do you change? Or close-source 🤐

**Comment by benhillis** [comment]
The change is in our init daemon which is closed-source, maybe someday we can open it up.  I can describe the change at a high level though:

```
1. Call getpwuid for default UID, if success goto 4
2. Call getpwuid for root UID, if success goto 4
3. Initialize a default passwd structure with some sane defaults (root, /bin/sh, etc).
4. Launch shell from passwd structure
```

**Comment by Brian-Perkins** [comment]
This change is now available in the Skip-ahead [17627](https://github.com/MicrosoftDocs/WSL/releases/tag/17627) build.

**Comment by Biswa96** [comment]
`wsl.exe` can't open without `passwd` but `bash.exe` can. Is it normal?

**Comment by minecraftmasterfun** [comment]
I had a similar issue. I set it to root to reset my Kali password, then I went to set it back. I ran cat /etc/passwd to get my user ID, then I went to the registry editor and changed the default UID to 1000. Of course, it didn't work, and I got error 0x80040305. So I went back, switched it to decimal, not hexadecimal 1000, and it worked. I know hex 0 and decimal 0 are the same but that might fix it.
